### ISSUE TITLE
Prediction order

### ISSUE THREAD

**Issue by adeldaoud** [grf-labs/grf#1084]
After fitting a causal forest, I would like to produce tau predictions for each unit and attach it back to the original data to probe CATE. I was under the impression that the order of units (rows) is preserved when executing the predict function like so `predict(tau.forest)$predictions`. However, it looks like this impression is not correct, as the following test shows. 

First, we generate data. Second, we execute `predict(tau.forest)$predictions`. Third, we compare these predictions by giving `predict()` the exact same dataset explicity. The order of this two prediction should be the same, - that is they should have the same tau values--- but they do not have the same order. I have `set.seed()` to make sure that any random draws are the same.

Is this a bug, or am I missing something?

```
library(grf)

set.seed(5)
n <- 2000
p <- 10
X <- matrix(rnorm(n * p), n, p)


W <- rbinom(n, 1, 0.4 + 0.2 * (X[, 1] > 0))
Y <- pmax(X[, 1], 0) * W + X[, 2] + pmin(X[, 3], 0) + rnorm(n)
tau.forest <- causal_forest(X, Y, W)
tau.forest

tau.hat.oob <- predict(tau.forest)$predictions
tau.hat.oob.test <- predict(tau.forest, X)$predictions

# should be the same values between the two vectors
tau.hat.oob[1:10]
tau.hat.oob.test[1:10]
```

**GRF version**
I am using GRF 2.02.

**Comment by erikcs** [comment]
Hi @adeldaoud, `predict(forest)` and `predict(forest, X)` do two different things. The first one gives "out of bag" (OOB) predictions using only trees that were not drawn for sample j, while the second predicts at sample j using all trees in the forest:

https://grf-labs.github.io/grf/reference/predict.causal_forest.html

>newdata | Points at which predictions should be made. If NULL, makes out-of-bag predictions on the training set instead (i.e., provides predictions at Xi using only trees that did not use the i-th training example).

OOB is a very convenient property in forest based estimators and is why you see statements like `Y - predict(forest)$predictions)` in the grf codebase: since the RHS is OOB we are not using information from sample j when subtracting its conditional mean estimate (which a traditional estimator would do with cross-fitting)

**Comment by adeldaoud** [comment]
@erikcs thanks for the clarification. OOB is convenient, and also faster. Then, a follow-up question is, how do I retain (or reorder) the OOB predictions `predict(forest)` to match the order of the original dataframe `X`? Basically, what I wish to do, is to attach the `predict(forest)$predictions)` back to the data, for further CATE probing. As specified in my code above, I would like to then do the following `X$tau <- predict(forest)` and then condition `X$tau` on columns of `X[]`.

**Comment by erikcs** [comment]
Hi @adeldaoud, the order of `predict(forsest)` will be in the same order as the original training data `X`

**Comment by adeldaoud** [comment]
@erikcs oh now I see. So the order is always the same for both `predict(tau.forest, X)$predictions` and `predict(tau.forest)$predictions` but the values differ because the former prediction uses OOB and the latter prediction uses all trees.
